opacclient/opacapp/src/main/java/de/geeksfactory/opacclient/reminder/SyncAccountService.java

```diff
@@ -0,0 +1,99 @@
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package de.geeksfactory.opacclient.reminder;
+
+import com.commonsware.cwac.wakeful.WakefulIntentService;
+
+import org.json.JSONException;
+
+import android.content.Context;
+import android.content.Intent;
+import android.content.SharedPreferences;
+import android.net.ConnectivityManager;
+import android.net.NetworkInfo;
+import android.preference.PreferenceManager;
+import android.util.Log;
+
+import java.io.IOException;
+import java.util.List;
+
+import de.geeksfactory.opacclient.BuildConfig;
+import de.geeksfactory.opacclient.OpacClient;
+import de.geeksfactory.opacclient.apis.OpacApi;
+import de.geeksfactory.opacclient.objects.Account;
+import de.geeksfactory.opacclient.objects.AccountData;
+import de.geeksfactory.opacclient.objects.Library;
+import de.geeksfactory.opacclient.storage.AccountDataSource;
+
+public class SyncAccountService extends WakefulIntentService {
+
+    private static final String NAME = "SyncAccountService";
+
+    public SyncAccountService() {
+        super(NAME);
+    }
+
+    @Override
+    protected void doWakefulWork(Intent intent) {
+        if (BuildConfig.DEBUG) Log.i(NAME, "SyncAccountService started");
+
+        SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(this);
+
+        ConnectivityManager connMgr = (ConnectivityManager) getSystemService(
+                Context.CONNECTIVITY_SERVICE);
+        NetworkInfo networkInfo = connMgr.getActiveNetworkInfo();
+        if (networkInfo != null) {
+            if (!sp.getBoolean("notification_service_wifionly", false) ||
+                    networkInfo.getType() == ConnectivityManager.TYPE_WIFI ||
+                    networkInfo.getType() == ConnectivityManager.TYPE_ETHERNET) {
+                syncAccounts();
+            }
+        }
+
+        if (BuildConfig.DEBUG) Log.i(NAME, "SyncAccountService finished");
+    }
+
+    private void syncAccounts() {
+        OpacClient app = (OpacClient) getApplication();
+        AccountDataSource data = new AccountDataSource(this);
+        data.open();
+        List<Account> accounts = data.getAccountsWithPassword();
+
+        for (Account account : accounts) {
+            if (BuildConfig.DEBUG) Log.i(NAME, "Loading data for Account " + account.toString());
+
+            try {
+                Library library = app.getLibrary(account.getLibrary());
+                if (!library.isAccountSupported()) continue;
+                OpacApi api = app.getNewApi(library);
+                AccountData res = api.account(account);
+                if (res == null) continue;
+                data.storeCachedAccountData(account, res);
+                data.close();
```

Comment:
You are using ``data.close`` after every account but ``data.open`` only once. Do I miss something or will this fail with more than one account?

Also, we should not keep the DataSource open during network operations to prevent concurrency issues. Therefore, we should close it after line 76 and open it again just before line 87.